### ISSUE TITLE
Auto-type enters "^" incorrectly, sometimes

### ISSUE THREAD

**Issue by max-rosenfors** [keepassxreboot/keepassxc#7610]
## Overview

After the most recent update auto-type has begun behaving oddly with some characters, notably `^`.

It does not enter it correctly into any window all the time, neither browser or text editor. Instead, it got replaced with some other character, `&` in my case.

In this attached screenshot you see the auto-type sequence performed in text editor, a few times over because the problem was not always present.

![image](https://user-images.githubusercontent.com/16334548/159669235-b27d1bfb-c3dd-4b76-9fe0-8d1a6834929a.png)

*`*`-have been added to hide parts of the old password, which is now also changed on my system.*

## Steps to Reproduce

1. Run Ubuntu 20.04.4 LTS (default Gnome desktop)
2. Have a auto-type password with the character `^`. (Or possibly a username also, not tested.)
3. Auto-type the password into an editor a few times, it should fail to input `^` after a few attempts.
    - When I first tried to enter my credentials on my system and discovered this bug, it did not successfully enter it correctly for three times in a row, contrary to how it seems to mostly enter it correctly now.

## Expected Behavior

Auto-type should be able to enter any character, or warn that a password contains characters that may not be entered through auto-type.

## Actual Behavior

It failed on some attempts, but as noted:

> When I first tried to enter my credentials on my system and discovered this bug, it did not successfully enter it correctly for three times in a row, contrary to how it seems to mostly enter it correctly in the text editor.

## Context

A workaround was to copy & paste the correct credentials into my system. After that I changed the password to one that did not contain the character `^`.

I notice that typing the character `^` (`LShift` + `¨` on my keyboard) produces an underscore under the character. I'm not sure if it always did this on my desktop, but it seems vaguely familiar. After that I can press any character to combine the `^` with that character. Ex: `âĉÔî`. It is possible that this chord sequence is what is causing trouble. I updated a few different parts of my system this morning, and other parts have changed at the same time that KeepassXC was updated.

I also wish to thank the maintainers of this project for providing a great password manager for multiple OS:es. It has performed flawlessly up until now, which is a very high mark of quality.

KeePassXC - Version 2.7.0
Revision: 20c9ac1
Distribution: Snap

Qt 5.15.3
Debugging mode is disabled.

Operating system: Ubuntu Core 20
CPU architecture: x86_64
Kernel: linux 5.13.0-37-generic

Enabled extensions:
- Auto-Type
- Browser Integration
- SSH Agent
- KeeShare
- YubiKey
- Secret Service Integration

Cryptographic libraries:
- Botan 2.12.1

Operating System: Ubuntu Linux
Desktop Env: Gnome
Windowing System: X11

**Comment by hifi** [comment]
What is the exact keyboard layout you use? It's very likely as you guessed that the way the character is produced with your keyboard layout is timing sensitive.

If you increase the Auto-Type typing delay in _Tools -> Settings -> General -> Auto-Type -> Auto-Type typing delay_ does that make it work better?

**Comment by hifi** [comment]
I think what's happening (guessed de layout) is that ^ is going into the fallback of creating that character on-the-fly because the new code doesn't understand the sequence to create it with your current layout and the fallback isn't always working correctly.

Can you try adding a second layout - the basic US English - into your available keyboard layouts even though you don't use it as your primary layout? This should avoid the fallback and instead cause a layout switch inside the sequence.

There are two different bugs here, I'll see if I can fix the confusion about creating ^ but the failed fallback is a mystery to me if it in fact is the case.

**Comment by djgera** [comment]
Yes this is really ugly, but is a new feature for compatibility with keepass [#5864](https://github.com/keepassxreboot/keepassxc/pull/5864) for + ^ and %

Use {^} instead.

**Comment by hifi** [comment]
> Yes this is really ugly, but is a new feature for compatibility with keepass [#5864](https://github.com/keepassxreboot/keepassxc/pull/5864) for + ^ and %
> 
> Use {^} instead.

No, this is about the password or username containing characters that are dead keys on your keyboard layout.

**Comment by max-rosenfors** [comment]
> What is the exact keyboard layout you use?

Swedish. Following is the output of `localectl status`:

```
$ localectl status
   System Locale: LANG=en_US.UTF-8
                  LC_NUMERIC=sv_SE.UTF-8
                  LC_TIME=sv_SE.UTF-8
                  LC_MONETARY=sv_SE.UTF-8
                  LC_PAPER=sv_SE.UTF-8
                  LC_NAME=sv_SE.UTF-8
                  LC_ADDRESS=sv_SE.UTF-8
                  LC_TELEPHONE=sv_SE.UTF-8
                  LC_MEASUREMENT=sv_SE.UTF-8
                  LC_IDENTIFICATION=sv_SE.UTF-8
       VC Keymap: n/a
      X11 Layout: se
       X11 Model: pc105
```

> Can you try adding a second layout - the basic US English

I went ahead and added English (US) as a second keyboard layout and switched to it, to test if that gave different results. However, now after lunch testing this (and a unrelated reboot) I am unable to reproduce the above results which I opened the issue with. Neither the Swedish layout or the English (US) layout have problems outputting the `^`-character. I made sure to use the same sequence of letters at the end there, as the start, and attempting to input this sequence works fine: `testuser	testpasswordci^0&W`.

I'm beginning to think the problem may have been completely unrelated to KeepassXC, and that I this morning managed to enter some wierd chord-mode that messes with it for more or less natural reasons. And that this semi chord-mode somehow kept itself active until i rebooted the computer.

The only other setting that I remember ever messing around with regarding keyboard layout and chords is adding this alias to my .bashrc: `alias code='GTK_IM_MODULE="xim" code'`. [It's because VS Code uses `Ctrl` + `Shift` + `e` as a combination internally](https://askubuntu.com/questions/1166660/how-to-add-open-vscode-with-gtk-im-module-xim), but it is also a system wide shortcut to enter a chord mode for inputting emoji. Adding that alias should disable that system-wide chord-mode shortcut when inside VS Code.

-----

I re-read your comment and noticed that you mentioned that I should only add the English (US) as a secondary layout, and let it be switched to automatically rather than switching to it myself. When doing this I am also unable to reproduce the issue.

- Swedish layout only: unable to reproduce, small stutter when it auto-types the `^`.
- Swedish layout + English (US) secondary: unable to reproduce, small stutter when it auto-types the `^`.
- English (US) primary + Swedish secondary: unable to reproduce, __no stutter__ when it auto-types the `^`.

I'm a happy camper since it works, I suppose, but you know how things that go seemingly unresolved can still bother you. Perhaps the issue can be closed until I or someone else experiences the same again.

**Comment by hifi** [comment]
Thanks for detailed testing report!

I'm surprised it still stutters with both layouts existing as the stutter comes from key emulation and not layout switching - at least for me.

The PR I opened should remove the stutter with only Swedish layout enabled as it handles the compositing dead key for ^ that in 2.7.0 is ignored and falls back to creating a new key temporarily.

Let's keep this issue open for a while if someone else can reproduce the original issue.

**Comment by ghost** [comment]
It also happens that "\`" isn't typed at all
So if the password is: abcdefg\`hijkl\`mno\`pqrstuvw\`xyz
sometimes the auto type output is: abcdefghijkl\`mno\`pqrstuvw\`xyz or abcdefg\`hijklmno\`pqrstuvw\`xyz etc.

**Comment by ghost** [comment]
"^" works for me perfectly
my localectl status output is:

```
$ localectl status
   System Locale: LANG=en_US.UTF-8
       VC Keymap: de-latin1
      X11 Layout: n/a
```

**Comment by hifi** [comment]
@Patrick260 what desktop are you running? Do you get stutters when backtick is being written?

**Comment by ghost** [comment]
lxde and os is arch

> Do you get stutters when backtick is being written?

What exactly do you mean?

**Comment by hifi** [comment]
Like it pauses for a few milliseconds longer during typing the letter than others.

**Comment by ghost** [comment]
> Like it pauses for a few milliseconds longer during typing the letter than others.

No

**Comment by hifi** [comment]
@Patrick260 I believe your backtick issue is fixed by the same change that went into 2.7.1. Original issue is still unresolved, unfortunately, as I can't reproduce it.

**Comment by rbaeta** [comment]
Hey!

This also happens to me :)

Tested with:
username: `rui.baeta^test` 
password: `1234^5678`
auto-type sequence: `{DELAY 1000}{USERNAME}{ENTER}{PASSWORD}{ENTER}`

Results of applying global auto-type for 6 times with typing delay set to 50ms:
```
rui.baeta^test
1234^5678

rui.baeta^test
1234&5678

rui.baeta&test
1234^5678

rui.baeta^test
1234^5678

rui.baeta^test
1234^5678

rui.baeta^test
1234&5678
```

Same test with typing delay set to 150ms:
```
rui.baeta^test
1234&5678

rui.baeta^test
1234&5678

rui.baeta^test
1234&5678

rui.baeta^test
1234^5678

rui.baeta^test
1234&5678

rui.baeta^test
1234&5678
```

As you can see, sometimes the char `^` is replaced by `&`. The typing delay seems not related to the issue.

Keepassxc 2.7.4
Keyboard layout: portuguese
System: Ubuntu 18.04.6 LTS
Linux laptop-1234 4.15.0-200-generic #211-Ubuntu SMP Thu Nov 24 18:16:04 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux

If you need help to debug this issue feel free to ask!

Thank you

**Comment by wesnm** [comment]
I've just encountered something similar, except with the '<' character being replaced by '\' (backslash) and '>' replaced by '|' (pipe).

Keepassxc 2.7.1
Keyboard layout: US English 105key
locate en_US.UTF-8
XFCE 4.18
FreeBSD 13.1-STABLE

**Comment by jibalamy** [comment]
I noticed that my problem with "@", which seems to be the same as the one expressed here, was not present in KeepassXC 2.6.6. It appeared in 2.7.

**Comment by Chealer** [comment]
@max-rosenfors: Are you convinced that this is specific to Ubuntu 20.04? If so, do you mind if we dismiss this report, since that is outdated?
If not, does this persist in 2.7.12 or ulterior? And if so, do you get problems with "@"?
